toolkit/components/filewatcher/NativeFileWatcherMac.h

```diff
@@ -0,0 +1,255 @@
+#include "nsIObserver.h"
+#include "nsThreadUtils.h"
+#include <mozilla/Mutex.h>
+
+#include <CoreServices/Components.k.h>
+#include <CoreServices/CoreServices.h>
+
+#include <queue>
+#include <vector>
+#include <sys/syslimits.h>
+
+
+namespace mozilla {
+
+namespace moz_filewatcher{
+
+/**
+ * A structure to hold the information about a single watched resource.
+ */
+struct WatchedResourceDescriptor
+{
+    // The path on the file system of the watched resource.
+    nsString mPath;
+    WatchedResourceDescriptor(const nsAString& aPath)
+        : mPath(aPath)
+    {
+    }
+};
+
+/**
+ * The CallBackAction struct is a specific event that occurs in the filesystem as reported by FSEvents.
+ */
+struct CallBackAction{
+    CallBackAction(char*eventPath, const FSEventStreamEventFlags eventFlag, const FSEventStreamEventId eventIds)
+        : mEventFlags(eventFlag)
+        , mEventIds(eventIds)
+    {
+        snprintf(mEventPath, PATH_MAX, "%s", eventPath);
+    }
+
+    char mEventPath[PATH_MAX];
+    const FSEventStreamEventFlags mEventFlags;
+    const FSEventStreamEventId mEventIds;
+};
+
+
+/**
+ * The CallBackEvents struct manages the FSEvents events returned to the stream.
+ *
+ * This is populated inside of the FSEvents callback. It is then processed on a separate
+ * thread by NativeFileWatcherIOTask::RunInternal() which parses the events and calls the
+ * appropriate callbacks.
+ */
+struct CallBackEvents{
+    CallBackEvents()
+        : callBackLock("filewatcher::eventLock")
+    {}
+    std::queue<CallBackAction> mSavedEvents; // Used to hold individual FSEvents
+    Mutex callBackLock; // Used for safe access across multiple threads.
+};
+
+class NativeFileWatcherIOTask; // Forward Declaration of class.
+
+
+/**
+ * The NativeFileWatcherFSETask class which holds all of the functionality of the FSEventStream.
+ *
+ * It inherits from Runnable as we need to make a blocking call to CFRunLoopRun(). CFRunLoopRun() allows the
+ * stream to report file system events. The event stream is an immutable object, therefore with every change in the
+ * paths watched, the stream must be stopped and a new stream must be created with a new list of paths to be watched.
+ */
+class NativeFileWatcherFSETask : public Runnable
+{
+public:
+    explicit NativeFileWatcherFSETask(NativeFileWatcherIOTask* parent, CallBackEvents* cbe, std::vector<CFStringRef>& dirs);
+
+    // This is a blocking method which runs the steam on a CFRunLoop.
+    NS_IMETHOD Run() override;
+
+    // Adds path to the stream (and restarts it)
+    NS_IMETHOD AddPath(char* pathToAdd);
+
+    // Removes path from the stream and restarts it. Unless it removes the last path being watched by the stream.
+    // In this instance, we will return from this runnable without starting a new stream.
+    NS_IMETHOD RemovePath(char* pathToRemove);
+
+private:
+    //mozilla::Mutex runLoopLock;
+
+    FSEventStreamRef mEventStreamRef; // Reference to the FSEventStream

```

Comment:
That's not a very useful comment. Could you rather document what the `FSEventStream` is for and who owns it?

Also, could you move comments above the declaration, rather than after, here and everywhere else?

Reply:
@fennnn to review